nbrowser

```diff
@@ -296,39 +310,50 @@ main(){
 		# bangs
 		"!"*)
 			local bang="${1/?}"
+			shift
 			if [ -f "${NBROWSER_CONFIG_DIR}/bangs/${bang}" ] ; then
-				shift
 				source "${NBROWSER_CONFIG_DIR}/bangs/${bang}"
 				has nbrowser_bang || _pemx "couldn't find nbrowser_bang() function in ${NBROWSER_CONFIG_DIR}/bangs/${bang}"
 				nbrowser_bang "$*"
 			else
-				_pemx "bang '${bang}' not found!"
+				bang=$(listfiles bangs | rofi -dmenu -p 'bangs' -i -mesg "bang '${bang}' not found!")
+				[ -z "${bang}" ] && exit
+				source "${NBROWSER_CONFIG_DIR}/bangs/${bang}"
+				has nbrowser_bang || _pemx "couldn't find nbrowser_bang() function in ${NBROWSER_CONFIG_DIR}/bangs/${bang}"
+				nbrowser_bang "$*"
+				#_pemx "bang '${bang}' not found!"
 			fi
 			;;
 
 		# engines
 		"?"*)
 			local engine="${1/?}"
+			shift
 			if [ -f "${NBROWSER_CONFIG_DIR}/engines/${engine}" ] ; then
-				shift
 				source "${NBROWSER_CONFIG_DIR}/engines/${engine}"
 				has nbrowser_search || _pemx "couldn't find nbrowser_search() function in ${NBROWSER_CONFIG_DIR}/engines/${engine}"
 				[ -n "$1" ] && url_handler "$(nbrowser_search $*)"
 			elif [ -n "${ENGINES[$engine]}" ]; then
-				shift
 				url_handler "${ENGINES[$engine]}$*"
-			elif [ -f "$NBROWSER_CONFIG_DIR/engines/engines.json" ]; then
-				shift
-				local url=$(jq -r ".[]|select( .t == \"$engine\" )|.u" "$NBROWSER_CONFIG_DIR/engines/engines.json" | sed "s/{{{s}}}/$*/g")
-				[ -z "$url" ] && _pemx "engine '${engine}' not found!"
-				url_handler "$url"
+			elif [ -f "$NBROWSER_CONFIG_DIR/engines.json" ]; then
+				local url=$(jq -r ".[]|select( .t == \"$engine\" )|.u" "$NBROWSER_CONFIG_DIR/engines.json" | sed "s/{{{s}}}/$*/g")
+				if [ ! -z "$url" ] ; then
+					url_handler "$url"
+				else
+					_pemx "engine '${engine}' not found!"
```

Comment:
instead of showing error, list all available engines to chose from
- include engines.json
- $ENGINES